src/Helpers/ApartmentHelpers.php

```
@@ -0,0 +1,44 @@
+<?php
+
+namespace BuildEmpire\Apartment\Helpers;
+
+class ApartmentHelpers
+{
+    /**
+     * Does the schema only contain letters and/or numbers.
+     *
+     * @param $schemaName
+     * @return bool
+     */
+    public static function isSchemaNameValid($schemaName)
+    {
+        return (preg_match('/^[a-z][a-z0-9_]*$/', $schemaName) && !ctype_upper($schemaName));
+    }
+
+    /**
+     * This will remove any bad characters from the $schemaName variable.
+     *
+     * Note: Removes any non valid strings with #. You should be checking with isSchemaNameValid before reaching this
+     * section. This is just an extra layer of protection.
+     *
+     * @param $schemaName
+     * @return mixed
+     */
+    public static function getSchemaSafeString($schemaName) {
+        return preg_replace('/^[^0-9a-z_]+$/', '#',$schemaName);
```

Any reason we're replacing them with hash? Is `#` within a schema name a sensible idea?

---

Some of the statements like CREATE SCHEMA and prefixing tables will not go through the normal PDO way.

It's a last minute attempt to secure the sql string before it goes out. All the methods should throw an exception if it's not a valid schema, but as an extract precaution this method is applied just before DB call.

At first I removed the invalid character, but what if this happened:

Good guy:
Company BuildEmpire
Schema: buildempire

Bad Guy:
JonDelHall
Schema: buildempire^

If the end-user was not responsible for removing/checking the schemas he/she is creating then: 

In this case the getSchemaSafeString would return:
buildempire#

In the old case it would return:
buildempire

which would be harmful.

---

And, of course, the raw sql statements are behind methods that validate the schema names.

---

Yea, but my thing is -- if it did fall through, the schema name would fo from "JonDellHall" - to be `#en#el#all` - not ideal.

Not even sure if you're able to use a hash in a schema name?

---

No, it would fail. It would throw an statement exception. I agree, this is not an ideal solution but the reason for doing this really are:

1. Anyone writing/extending Apartment in the future should be checking the data passing in is correct, which all our methods should be doing.

2. It's a last sort of attempt to stop anything bad being passed into the statement.

It's not really required as we are checking the data before it's passed but just an extra layer of security.

---

I don't like this ... not sure it's a suitable solution. Deliberately putting breaking code in as a security method is not right. 

Not happy with this, sorry.

---

yeah, i agree. I've thought of a different approach. Pretty terrible idea sorry will fix